### ISSUE TITLE
changed source while building, and bazel got into stable inconsistent state

### ISSUE THREAD

**Issue by mafanasyev-tri** [bazelbuild/bazel#4331]
### Description of the problem / feature request / question:

While I was building our project using bazel  (with  `bazel test //..`"), I accidentally checked out a different revision of the code. When I discovered it, I restored the previous revision, stopped bazel mid-build, and then restarted to get back to the consistent state. However, one of the unit tests has failed.

I re-ran the bazel it a couple of times, and nothing changed -- bazel did not detect that any file needed rebuilding, and the same error kept happening again and again. Running "bazel clean" followed by another rebuild cleared the error -- the test started to pass.

The failing test was a C++ app. It looks like the problem occurred because two object files from different revisions got linked together -- so a test driver was from one revision, white the module under test was from the old one.

This looks like a classical example of "stable inconsistent state".  The user manual says:

> If you ever detect a stable inconsistent state with Bazel, please report a bug.

so I am reporting this bug. But the same manual says that only _normal or "reasonable" use of the build tool_ counts, and I am not sure if this situation counts as reasonable.

### If possible, provide a minimal example to reproduce the problem:

I am not sure how to investigate this. I have saved bazel caches from inconsistent run and from successful rebuild. I tried diff'ing them, and there are too many differences to find meaningful changes. I am not sure what else would I be looking for.

If this bug is valid, I can try to set up the reproduction.

### Environment info
* Operating System:

Ubuntu 16.04.3 LTS xenial
4.4.0-97-generic

* Bazel version (output of `bazel info release`):

Build label: 0.8.0- (@non-git)

* If `bazel info release` returns "development version" or "(@non-git)", please tell us what source tree you compiled Bazel from; git commit hash is appreciated (`git rev-parse HEAD`):

0.8.0 source zip from releases page

###  Have you found anything relevant by searching the web?
searched the bug tracker and bazel-discuss for "inconsistency" and "inconsistent", found nothing relevant

### Anything else, information or logs or outputs that would be helpful?

bazel was running in `--batch` mode, in the docker container

**Comment by aehlig** [comment]
It is definitely a "reasonable" use of a build tool, to interrupt it. If a build tools does not clean up enough (invalidating too much would be OK, as interruption is not the use case to optimize for) after being interrupted that certainly is a bug and we have to take it very seriously.

As it happened during a C++ build, I'm asking our C++ experts to have a look first, even though it might be a more fundamental issue.

/cc @ulfjack @lberki

**Comment by mhlopko** [comment]
I don't think this is C++ specific, more of an action cache race. Pinging @ericfelly.

**Comment by ulfjack** [comment]
This is probably the same as #3360. We believe the same race condition also applies to the local action cache.

**Comment by ulfjack** [comment]
I have a fix for #3360, but it is not integrated with the local action cache yet. I'll keep this open with the intent of also fixing the local action cache.
